tools/toolchains/iar.py

```diff
@@ -55,7 +55,7 @@ def __init__(self, target, notify=None, macros=None, build_profile=None,
             build_profile=build_profile
         )
         core = target.core
-        if CORE_ARCH[target.core] == 8:
+        if CORE_ARCH[target.core] == 8 and "TFM" in target.labels:
```

Comment:
If testing TFM, I would take out the `CORE_ARCH` test

Reply:
please consider following approach:
```suggestion
        if CORE_ARCH[target.core] == 8 and target.is_PSA_secure_target:
```
We are only injecting this define to a linker script. Please consider adding this macro also for compile flags for consistency.

Reply:
What I'd like to see is a core-independent "we're making a TrustZone/CMSE-type split image" test here.

If I understand correctly, `is_PSA_secure_target` covers TrustZone and dual-core solutions, and there's no fundamental reason why someone might not end up making a dual-M33 device.

So "TFM" is closer to the meaning we want. In #9460 I suggested adding `Target.is_TZ_secure_target` and `Target.is_TZ_nonsecure_target` for this bit of code.

If not changing anything else, those could be defined as `'TFM' in labels and [not] self.core.ends_with('-NS') `

Although for all of the three possibilities suggested here, we're not covering "NUMAKER_PFM_M2351", which doesn't have TFM/PSA.

But then that's why @bridadan was looking at adding a separate explicit `tz` flag.

Reply:
@alzix proposition is fine for me

@kjbracey-arm 

> we're not covering "NUMAKER_PFM_M2351"

For me, there is no issue as NUMAKER_PFM_M2351 is Cortex-M23-NS

Reply:
is

if CORE_ARCH[target.core] == 8 and (target.is_PSA_secure_target or "MBED_TZ_DEFAULT_ACCESS=1" in target.macros):

would be fine for all ?

Reply:
I really don't want `CORE_ARCH` in there, on principle. The tying of security to core version is the whole problem we're trying to fix, so even if this is a temporary "heuristic" until a more explicit flag is added for targets.json, I want that to be an architecture-free heuristic if at all possible.

We can cover M2351's use of NS by just detecting the "-NS" suffix on the core, without caring about any other settings.

I believe it should be flattened out, as per my initial comment on #9460, so it's something like

    if `TFM` in target.labels:
        if not target.core.endswith("-NS"):
            # XXX Why is this conditional? Don't we always create secure library
            # if setting secure build flag?
            if (kwargs.get('build_dir', False)):  
                # Create Secure library

            # Add secure build flag
 
    if target.core.ends_with('-NS'):
        # Add linking time preprocessor macro DOMAIN_NS

I didn't quite get @alzix's objection to looking at label `TFM` rather than `SPE_Target` or `NSPE_Target`.

Reply:
Oh, and I still don't think there's any need to be explicitly setting DOMAIN_NS to 0, but let me know if I'm wrong.